// reactour/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function get() {
    return _components.Arrow;
  }
});
Object.defineProperty(exports, "Close", {
  enumerable: true,
  get: function get() {
    return _components.Close;
  }
});
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function get() {
    return _components.Badge;
  }
});
Object.defineProperty(exports, "Controls", {
  enumerable: true,
  get: function get() {
    return _components.Controls;
  }
});
Object.defineProperty(exports, "Navigation", {
  enumerable: true,
  get: function get() {
    return _components.Navigation;
  }
});
Object.defineProperty(exports, "Dot", {
  enumerable: true,
  get: function get() {
    return _components.Dot;
  }
});
exports.default = void 0;

var _Tour = _interopRequireDefault(require("./Tour"));

var _components = require("./components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _Tour.default;
exports.default = _default;
